# One sample t-test

##1.one sample t-test

* 20대 평균 수면시간에 대한 one sample t-test
  * 유의수준은 5% 사용
  

In [3]:
#[0] 수면 시간 정보가 포함된 파일 불러오기
import pandas as pd
import numpy as np
data = {
    "stime20s": [4, 4, 5, 5, 6, 6, 7, 3, 8, 7, 8, 8, 7, 7, 7, 6, 8, 5, 6, 6],
    "stime40s": [5, 5, 6, 7, 6, 7, 7, 5, 5, 6, 8, 7, 6, 6, 8, 7, 7, 6, 5, 8],
    "ID": range(1, 21)
}

df = pd.DataFrame(data)
print(df.shape)
df.T

(20, 3)


0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  \
stime20s   4   4   5   5   6   6   7   3   8   7   8   8   7   7   7   6   8   
stime40s   5   5   6   7   6   7   7   5   5   6   8   7   6   6   8   7   7   
ID         1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17   

          17  18  19  
stime20s   5   6   6  
stime40s   6   5   8  
ID        18  19  20

In [4]:
# [1] 20대의 수면시간에 대해 정규성 검정
# one sample이므로 등분산 검정 필요 없음
from scipy.stats import shapiro
statistic,pvalue = shapiro(df['stime20s'])
print(round(statistic,4),round(pvalue,4))
print('기각 ' if pvalue<0.05 else '채택')

0.9239 0.118
채택


In [5]:
#[2] 20대의 수면시간에 대한 평균 구하기
sleep = df['stime20s'].mean()
print(round(sleep, 4), f'{sleep:.4f}')

6.15 6.1500


##2.가설 1. alternative = 'two-sided'


* 가설 1. alternative = 'two-sided'
  * 귀무가설 : 20대 수면시간의 평균은 6시간이다
  * 대립가설 : 20대 수면시간의 평균은 6시간이 아니다

In [6]:
# [3] One Sample t-test 수행
from scipy.stats import ttest_1samp
popmean = 6
statistic,pvalue = ttest_1samp(df['stime20s'],
                               popmean,alternative = 'two-sided')
print(round(statistic,4),round(pvalue,4))

# [4] 가설 결과
# p-value가 0.05보다 크다
# 귀무가설을 채택해야 한다
# 20대 수면시간의 평균은 6시간이다
print('기각' if pvalue<0.05 else '채택')

0.4592 0.6513
채택


##3.가설 2. alternative = 'less'


* 가설 2. alternative = 'less'
  * 귀무가설 : 20대 수면시간의 평균은 6시간 보다 크거나 같다.
  * 대립가설  : 20대 수면시간의 평균은 6시간 보다 작다.


In [7]:
# [5] One Sample t-test수행
from scipy.stats import ttest_1samp
statistic,pvalue = ttest_1samp(df['stime20s'],
                               popmean, alternative='less')
print(round(statistic,4),round(pvalue,4))

#[6] 가설 결과
# p-value가 0.05보다 크다
# 귀무가설을 채택해야한다
# 20대 수면시간의 평균은 6시간 보다 크거나 같다.
print('기각' if pvalue<0.05 else '채택')

0.4592 0.6743
채택


##4.가설 3. alternative = 'greater'


* 가설 3. alternative = 'greater'
  * 귀무가설 : 20대 수면시간의 평균은 6시간보다 작거나 같다.
  * 대립가설 : 20대 수면시간의 평균은 6시간 보다 크다
  

In [8]:
# [7] One Sample t-test 수행
from scipy.stats import shapiro
statistic,pvalue = ttest_1samp(df['stime20s'],
                               popmean,alternative='greater')
print(round(statistic,4),round(pvalue,4))

# [8] 가설 결과
# pvalue 가 0.05보다 크다
# 귀무가설을 채택해야한다
# 20대 수면시간의 평균은 6시간보다 작거나 같아
print( '기각' if pvalue<0.05 else '채택')

0.4592 0.3257
채택


In [25]:
from scipy.stats import t, sem

target = df['stime20s']
# 자유도, 평균, 표준오차
df_value = len(target) - 1
mean = target.mean()
std_err = sem(target)

# 신뢰구간 계산
lower = mean + t.ppf(0.025, df_value) * std_err
upper = mean + t.ppf(0.975, df_value) * std_err

print(f'{lower:.4f} ~ {upper:.4f}')


5.4663 ~ 6.8337


In [24]:
from scipy.stats import t, sem

target = df['stime20s']
# t-분포 객체를 생성하면서 loc와 scale을 설정
t_dist = t(df=len(target)-1, loc=target.mean(), scale=sem(target))
lower, upper = t_dist.interval(0.95)  # 신뢰 구간을 계산

print(f'{lower:.4f} ~ {upper:.4f}')


5.4663 ~ 6.8337


In [37]:
from scipy.stats import t,sem

target = df['stime20s']
# t-분포 객체를 생성하면서 loc와 scale을 설정
t_dist = t(df=len(target)-1,loc = target.mean(), scale = sem(target))
lower,upper = t_dist.interval(0.95) # 신뢰구간을 계산
print(f'{lower:.4f} ~ {upper:.4f}')


5.4663 ~ 6.8337


# Two sample t-test


##1.'연령에 따른 수면 시간 two sample t-test 유의수준은 5% 사용


In [40]:
#[1] 수면 시간 정보가 포함된 파일 불러오기
import pandas as pd
import numpy as np
data = {
    "stime20s": [4, 4, 5, 5, 6, 6, 7, 3, 8, 7, 8, 8, 7, 7, 7, 6, 8, 5, 6, 6],
    "stime40s": [5, 5, 6, 7, 6, 7, 7, 5, 5, 6, 8, 7, 6, 6, 8, 7, 7, 6, 5, 8],
    "ID": range(1, 21)
}

df2 = pd.DataFrame(data)
print(df2.shape)
df2.T

(20, 3)


0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  \
stime20s   4   4   5   5   6   6   7   3   8   7   8   8   7   7   7   6   8   
stime40s   5   5   6   7   6   7   7   5   5   6   8   7   6   6   8   7   7   
ID         1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17   

          17  18  19  
stime20s   5   6   6  
stime40s   6   5   8  
ID        18  19  20

In [52]:
#[2] 그룹 분할하기
groupA = df2.loc[:,'stime20s']
groupB = df2.loc[:,'stime40s']

#[3] 그롭별 평균군하기
print(round(groupA.mean(),4),round(groupB.mean(),4))

6.15 6.35


In [60]:
#[4] 등분순 검정 - 3가지 방법으로 실행 후 ,pvalue 사용
from scipy.stats import levene,fligner,bartlett
from scipy.stats import shapiro

#정규성 검정
statistic,pvalueA = shapiro(groupA)
statistic,pvaleuB = shapiro(groupB)
print(f'{pvalueA:.4f},{pvaleuB:.4f}')

#등분산 검정
statisticA,pvalueA = levene(groupA,groupB)
statisticB,pvalueB = fligner(groupA,groupB)
statisticC,pvalueC = bartlett(groupA,groupB)
print(f'{pvalueA:.4f},{pvalueB:.4f},{pvalueC:.4f}')

0.1180,0.0184
0.2312,0.1808,0.1475


* Levene’s Test (p-value = 0.2312):

  * p-value가 0.05보다 크므로, 귀무가설을 기각할 수 없습니다.
  따라서, groupA와 groupB는 등분산을 만족한다고 결론 내릴 수 있습니다.
  Fligner-Killeen Test (p-value = 0.1808):

  * p-value가 0.05보다 크므로, 귀무가설을 기각할 수 없습니다.
  따라서, 이 검정에서도 groupA와 groupB는 등분산을 만족한다고 결론 내릴 수 있습니다.
  Bartlett's Test (p-value = 0.1475):

  * p-value가 0.05보다 크므로, 귀무가설을 기각할 수 없습니다.
  이 검정에서도 groupA와 groupB는 등분산을 만족한다고 결론 내릴 수 있습니다.

##2.가설 1. alternative = 'two sided'

* 귀무가설 : groupA의 평균 - groupB의 평균이 0과 같다
* 대립가설 : groupA의 평균 - groupB의 평균이 0과 같지 않다
* 귀무가설 : groupA,groupB의 평균은 동일하다.
* 대립가설 : groupA,groupB의 평균은 동일하지 않다

In [67]:
#[5] two-sample t-test 수행
from scipy.stats import ttest_ind
statistic,pvalue = ttest_ind(groupA,groupB,alternative='two-sided')
print(f'statistic : {statistic:.4f}, pvalue : {pvalue:.4f}')

#[6] 가설결과d
# pvalue가 0.05보다 큰값(0.6208)
# 귀무가설을 채택해야하 함

statistic : -0.4988, pvalue : 0.6208


##3.가설 2. alternative = 'less'


* 귀무가설 : groupA의 평균 - groupB의 평균이 ㅇ보다 크거나 작다
* 대립가설 : groupA의 평균 groupB의 평균이 0보다 작다.
* 귀무가설 : groupA의 평균이 groupB의 평균보다 크거나 같다
* 대립가설 : groupA의 평균이 groupB의 평균보다 작다

In [71]:
# [7] two-sample t-test 수행
from scipy.stats import ttest_ind
statistic,pvalue = ttest_ind(groupA,groupB,alternative='less')
print(f'statistic : {statistic:.4f}, pvalue : {pvalue:.4f}')

# [8] 가설 결과
# 귀무가설을 채택해야 함
# 즉 , groupA의 평균 - groupB의 평균이 0보다 크거나 작다
# groupA의 평균이 groupB의 평균보다 크거나 같다
print('기각' if pvalue<0.05 else '채택')

statistic : -0.4988, pvalue : 0.3104
채택


##4.가설 3. alternative = 'greater'



* 귀무가설 : groupA의 평균 - groupB의 평균이 0보다 작거나 같다
* 대립가설 : groupA의 평균 - groupB의 평균이 0보다 크다
* 귀무가설 : groupA의 평균이 groupB의 평균보다 작거나 같다
* 대립가설 : groupA의 평균이 groupB의 평균보다 크다

In [74]:
#[9] two-sample t-test 수행
from scipy.stats import ttest_ind
statistic,pvalue = ttest_ind(groupA,groupB,alternative = 'greater')
print(f'statistic : {statistic:.4f}, pvalue : {pvalue:.4f}')

#[10] 가설 결과
# pvalue가 0.05보다 큼
# 귀무가설을 채택함
# 즉,groupA의 평균 - groupB의 평균이 0보다 작거나 같다
# groupA의 평균이 groupB의 평균보다 작거나 같다
print('기각' if pvalue<0.05 else '채택')

statistic : -0.4988, pvalue : 0.6896
채택
